flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobRecoverableWriter.java

```diff
@@ -0,0 +1,103 @@
+import org.apache.flink.runtime.fs.hdfs.HadoopRecoverableSerializer;
+import org.apache.flink.runtime.fs.hdfs.HadoopRecoverableWriter;
+
+import org.apache.hadoop.fs.FileSystem;
+
+import java.io.IOException;
+
+/** Recoverable writer for AzureBlob file system. */
+public class AzureBlobRecoverableWriter extends HadoopRecoverableWriter {
+    /**
+     * Creates a new Recoverable writer.
+     *
+     * @param fs The AzureBlob file system on which the writer operates.
+     */
+    public AzureBlobRecoverableWriter(FileSystem fs) {
+        super(fs);
+    }
+
+    protected void checkSupportedFSSchemes(org.apache.hadoop.fs.FileSystem fs) {
+        // This writer is only supported on a subset of file systems
+        if (!("abfs".equalsIgnoreCase(fs.getScheme())
+                || "abfss".equalsIgnoreCase(fs.getScheme()))) {
+            throw new UnsupportedOperationException(
+                    "Recoverable writers on AzureBlob are only supported for ABFS");
+        }
+    }
+
+    @Override
+    protected RecoverableFsDataOutputStream getRecoverableFsDataOutputStream(
+            org.apache.hadoop.fs.Path targetFile, org.apache.hadoop.fs.Path tempFile)
+            throws IOException {
+        return new AzureBlobFsRecoverableDataOutputStream(fs, targetFile, tempFile);
+    }
+
+    @Override
+    public RecoverableFsDataOutputStream recover(ResumeRecoverable recoverable) throws IOException {
+        return new AzureBlobFsRecoverableDataOutputStream(fs, (HadoopFsRecoverable) recoverable);
+    }
+
+    @Override
+    public boolean requiresCleanupOfRecoverableState() {
+        return false;
+    }
+
+    @Override
+    public boolean cleanupRecoverableState(ResumeRecoverable resumable) throws IOException {
+        return false;
+    }
+
+    @Override
+    public RecoverableFsDataOutputStream.Committer recoverForCommit(CommitRecoverable recoverable)
+            throws IOException {
+        return new AzureBlobFsRecoverableDataOutputStream.ABFSCommitter(
+                fs, (HadoopFsRecoverable) recoverable);
+    }
+
+    @Override
+    public SimpleVersionedSerializer<CommitRecoverable> getCommitRecoverableSerializer() {
+        @SuppressWarnings("unchecked")
+        SimpleVersionedSerializer<CommitRecoverable> typedSerializer =
+                (SimpleVersionedSerializer<CommitRecoverable>)
+                        (SimpleVersionedSerializer<?>) HadoopRecoverableSerializer.INSTANCE;
+
+        return typedSerializer;
+    }
+
+    @Override
+    public SimpleVersionedSerializer<ResumeRecoverable> getResumeRecoverableSerializer() {
+        @SuppressWarnings("unchecked")
+        SimpleVersionedSerializer<ResumeRecoverable> typedSerializer =
+                (SimpleVersionedSerializer<ResumeRecoverable>)
+                        (SimpleVersionedSerializer<?>) HadoopRecoverableSerializer.INSTANCE;
+        return typedSerializer;
+    }
+
+    @Override
+    public boolean supportsResume() {
+        return true;
+    }
```

Comment:
```suggestion
```

Reply:
And then we won't need the visibility change for `HadoopRecoverableSerializer`.

Reply:
Got it sure.